coffee/factorizer.py

```diff
@@ -0,0 +1,233 @@
+#     * The name of Imperial College London or that of other
+#       contributors may not be used to endorse or promote products
+#       derived from this software without specific prior written
+#       permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTERS
+# ''AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+# OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import operator
+
+from base import *
+from utils import *
+
+
+class Term():
+    """A Term represents a product between 'operands' and 'factors'. In a
+    product /a*(b+c)/, /a/ is the 'operand', while /b/ and /c/ are the 'factors'.
+    The symbol /+/ is the 'op' of the Term.
+    """
+
+    def __init__(self, operands, factors=None, op=None):
+        self.operands = operands
+        self.factors = factors or []
+        self.op = op
+
+    @property
+    def operands_ast(self):
+        return ast_make_expr(Prod, self.operands)
+
+    @property
+    def factors_ast(self):
+        return ast_make_expr(self.op, self.factors)
+
+    @property
+    def generate_ast(self):
+        if len(self.factors) == 0:
+            return self.operands_ast
+        elif len(self.operands) == 0:
+            return self.factors_ast
+        elif len(self.factors) == 1 and \
+                all(isinstance(i, Symbol) and i.symbol == 1.0 for i in self.factors):
+            return self.operands_ast
+        else:
+            return Prod(self.operands_ast, self.factors_ast)
+
+    def add_operands(self, operands):
+        for o in operands:
+            if o not in self.operands:
+                self.operands.append(o)
+
+    def remove_operands(self, operands):
+        for o in operands:
+            if o in self.operands:
+                self.operands.remove(o)
+
+    def add_factors(self, factors):
+        for f in factors:
+            if f not in self.factors:
+                self.factors.append(f)
+
+    def remove_factors(self, factors):
+        for f in factors:
+            if f in self.factors:
+                self.factors.remove(f)
+
+    @staticmethod
+    def process(symbols, should_factorize, op=None):
+        operands = [s for s in symbols if should_factorize(s)]
+        factors = [s for s in symbols if not should_factorize(s)]
+        return Term(operands, factors, op)
+
+
+class Factorizer():
+
```

Comment:
just the main factorizer function (the only one the user is supposed to call, so what it does is basically described in the class' docstring)